src/goto-symex/field_sensitivity.cpp

```diff
@@ -0,0 +1,251 @@
+
+    if(
+      member.struct_op().id() == ID_symbol &&
+      member.struct_op().get_bool(ID_C_SSA_symbol) &&
+      ns.follow(member.struct_op().type()).id() == ID_struct)
+    {
+      // place the entire member expression, not just the struct operand, in an
+      // SSA expression
+      ssa_exprt tmp = to_ssa_expr(member.struct_op());
+      member.struct_op() = tmp.get_original_expr();
+      tmp.set_expression(member);
+
+      expr.swap(tmp);
+    }
+  }
+  else if(expr.id() == ID_index)
+  {
+    index_exprt &index = to_index_expr(expr);
+    simplify(index.index(), ns);
+
+    if(
+      index.array().id() == ID_symbol &&
+      index.array().get_bool(ID_C_SSA_symbol) &&
+      ns.follow(index.array().type()).id() == ID_array &&
+      index.index().id() == ID_constant)
+    {
+      // place the entire index expression, not just the array operand, in an
+      // SSA expression
+      ssa_exprt tmp = to_ssa_expr(index.array());
+      index.array() = tmp.get_original_expr();
+      tmp.set_expression(index);
+
+      expr.swap(tmp);
+    }
+  }
+#endif
+}
+
+exprt field_sensitivityt::get_fields(
+  const namespacet &ns,
+  const ssa_exprt &ssa_expr)
+{
+#ifdef ENABLE_FIELD_SENSITIVITY
+  const typet &followed_type = ns.follow(ssa_expr.type());
+
+  if(followed_type.id() == ID_struct)
+  {
+    const exprt &struct_op = ssa_expr.get_original_expr();
+
+    const struct_typet &type = to_struct_type(followed_type);
+
+    const struct_union_typet::componentst &components = type.components();
+
+    struct_exprt result(ssa_expr.type());
+    result.reserve_operands(components.size());
+
+    for(const auto &comp : components)
+    {
+      const member_exprt member(struct_op, comp.get_name(), comp.type());
+      ssa_exprt tmp = ssa_expr;
+      tmp.set_expression(member);
+      result.copy_to_operands(get_fields(ns, tmp));
+    }
+
+    return result;
+  }
+  else if(
+    followed_type.id() == ID_array &&
+    to_array_type(followed_type).size().id() == ID_constant)
+  {
+    const exprt &array = ssa_expr.get_original_expr();
+
+    const array_typet &type = to_array_type(followed_type);
+
+    const std::size_t array_size = numeric_cast_v<std::size_t>(type.size());
+
+    array_exprt result(type);
+    result.reserve_operands(array_size);
+
+    for(std::size_t i = 0; i < array_size; ++i)
+    {
+      const index_exprt index(array, from_integer(i, index_type()));
+      ssa_exprt tmp = ssa_expr;
+      tmp.set_expression(index);
+      result.copy_to_operands(get_fields(ns, tmp));
+    }
+
+    return result;
+  }
+  else
+#endif
+    return ssa_expr;
+}
+
+void field_sensitivityt::field_assignments(
+  const namespacet &ns,
+  goto_symex_statet &state,
+  symex_targett &target,
+  const exprt &lhs)
+{
+  exprt lhs_fs = lhs;
+  apply(ns, lhs_fs, false);
+
+  bool do_apply_bak = do_apply;
+  do_apply = false;
+  field_assignments_rec(ns, state, target, lhs_fs, lhs);
+  do_apply = do_apply_bak;
```

Comment:
The argument doesn't work, because we need to communicate this across calls to functions of `goto_symex_statet`. Instead, I have made the class stateful, and make sure that the full class is being passed around.